cloudinit/config/cc_puppet.py

```diff
@@ -344,7 +359,12 @@ def handle(
 
     if start_puppetd:
         # Start puppetd
-        subp.subp(["service", "puppet", "start"], capture=False)
+        for service_name in service_names:
+            try:
+                cloud.distro.manage_service("start", service_name)
+                break
+            except subp.ProcessExecutionError:
+                log.debug(f"Could not auto start {service_name}")
```

Comment:
Given that both the "enable" and "start" logic should probably be the same. It may be useful to generalize a function like _manage_one_of_puppet_services  to deal with both enable and start. Since we likely want to fallback to puppet if puppet-agent service isn't available, let's handle all fallback options in one place and only log.warning if we can't find a viable option


Here's a diff for you to consider if you think it is applicable/functional for this case. It let's us consolidate the _autostart_puppet for the "enable" case and the semi-duplication of the if start_puppetd checks below during "start"


```diff
diff --git a/cloudinit/config/cc_puppet.py b/cloudinit/config/cc_puppet.py
index d19eb43db..7cbe2c2e3 100644
--- a/cloudinit/config/cc_puppet.py
+++ b/cloudinit/config/cc_puppet.py
@@ -25,7 +25,7 @@ from cloudinit.settings import PER_INSTANCE
 
 AIO_INSTALL_URL = "https://raw.githubusercontent.com/puppetlabs/install-puppet/main/install.sh"  # noqa: E501
 PUPPET_AGENT_DEFAULT_ARGS = ["--test"]
-PREFERRED_PUPPET_PACKAGE_NAMES = ("puppet-agent", "puppet")
+PREFERRED_PUPPET_NAMES = ("puppet-agent", "puppet")
 
 MODULE_DESCRIPTION = """\
 This module handles puppet installation and configuration. If the ``puppet``
@@ -119,20 +119,6 @@ class PuppetConstants:
         self.csr_attributes_path = csr_attributes_path
 
 
-def _autostart_puppet(log, cloud: Cloud, service_name: str):
-    # Set puppet to automatically start
-    try:
-        if os.path.exists("/sbin/chkconfig"):
-            subp.subp(["/sbin/chkconfig", service_name, "on"], capture=False)
-        else:
-            cloud.distro.manage_service("enable", f"{service_name}.service")
-    except subp.ProcessExecutionError:
-        log.warning(
-            f"Sorry we do not know how to enable {service_name} services on \
-              this system"
-        )
-
-
 def get_config_value(puppet_bin, setting):
     """Get the config value for a given setting using `puppet config print`
     :param puppet_bin: path to puppet binary
@@ -178,6 +164,24 @@ def install_puppet_aio(
         return subp.subp([tmpf] + args, capture=False)
 
 
+def _manage_one_of_puppet_services(action: str):
+    """Attempt to perform action on one of puppet services."""
+    service_managed: str = ""
+    for puppet_name in PREFERRED_PUPPET_NAMES:
+        try:
+            cloud.distro.manage_service("enable", f"{puppet_name}.service")
+            service_managed = puppet_name
+            break
+        except subp.ProcessExecutionError:
+            pass
+    if not service_managed:
+        log.warning(
+            "Could not '%s' any of the following services: %s",
+            action,
+            ", ".join(PUPPET_PREFERRED_NAMES)
+        )
+
+
 def handle(
     name: str, cfg: Config, cloud: Cloud, log: Logger, args: list
 ) -> None:
@@ -218,7 +222,6 @@ def handle(
     package_name = util.get_cfg_option_str(
         puppet_cfg, "package_name", puppet_package
     )
-    service_names = ["puppet-agent"]
     if not install and version:
         log.warning(
             "Puppet install set to false but version supplied, doing nothing."
@@ -232,23 +235,17 @@ def handle(
 
         if install_type == "packages":
             if package_name is None:  # conf has no package_name
-                try:
-                    log.debug("Trying to install puppet-agent")
-                    cloud.distro.install_packages(
-                        (PREFERRED_PUPPET_PACKAGE_NAMES[0], version)
+                for puppet_name in PREFERRED_PUPPET_NAMES:
+                    try:
+                        cloud.distro.install_packages(puppet_name, version)
+                        package_name = puppet_name
+                    except subp.ProcessExecutionError:
+                        pass
+                if not package_name:
+                    log.warning(
+                        "No installable puppet package in any of: %s",
+                        ", ".join(PREFERRED_PUPPET_NAMES)
                     )
-                    service_names = ["puppet-agent"]
-                except subp.ProcessExecutionError:
-                    log.debug("Trying to install puppet")
-                    cloud.distro.install_packages(
-                        (PREFERRED_PUPPET_PACKAGE_NAMES[1], version)
-                    )
-                    service_names = ["puppet"]
-            else:
-                cloud.distro.install_packages((package_name, version))
-                # install both services to find the correct one
-                service_names = ["puppet-agent", "puppet"]
-
         elif install_type == "aio":
             install_puppet_aio(
                 cloud.distro, aio_install_url, version, collection, cleanup
@@ -326,14 +323,9 @@ def handle(
             yaml.dump(puppet_cfg["csr_attributes"], default_flow_style=False),
         )
 
-    # Set it up so it autostarts
     if start_puppetd:
-        for service_name in service_names:
-            try:
-                _autostart_puppet(log, cloud, service_name)
-                break
-            except subp.ProcessExecutionError:
-                log.debug(f"Could not auto enable {service_name}")
+        # Enable puppet auto-start on boot
+        _manage_one_of_puppet_services("enable")
 
     # Run the agent if needed
     if run:
@@ -358,13 +350,8 @@ def handle(
         subp.subp(cmd, capture=False)
 
     if start_puppetd:
-        # Start puppetd
-        for service_name in service_names:
-            try:
-                cloud.distro.manage_service("start", service_name)
-                break
-            except subp.ProcessExecutionError:
-                log.debug(f"Could not auto start {service_name}")
+        # Start puppetd now
+        _manage_one_of_puppet_services("start")
 
 
 # vi: ts=4 expandtab
```

Reply:
This is a goodt suggestion and improvement.